server/src/main/java/com/hedera/block/server/verification/session/AbstractBlockVerificationSession.java

```
@@ -0,0 +1,190 @@
+     */
+    protected final MetricsService metricsService;
+    /**
+     * The signature verifier.
+     */
+    protected final SignatureVerifier signatureVerifier;
+    /**
+     * The block number being verified.
+     */
+    protected final long blockNumber;
+    /**
+     * The tree hasher for input hashes.
+     */
+    protected final StreamingTreeHasher inputTreeHasher;
+    /**
+     * The tree hasher for output hashes.
+     */
+    protected final StreamingTreeHasher outputTreeHasher;
+    /**
+     * The time when block verification started.
+     */
+    protected final long blockWorkStartTime;
+    /**
+     * A flag indicating whether the session is running.
+     */
+    protected volatile boolean running = true;
+    /**
+     * The future for the verification result.
+     */
+    protected final CompletableFuture<VerificationResult> verificationResultFuture = new CompletableFuture<>();
+
+    /**
+     * Constructs the session with shared initialization logic.
+     *
+     * @param blockHeader the block header
+     * @param metricsService the metrics service
+     * @param signatureVerifier the signature verifier
+     * @param inputTreeHasher the input tree hasher (e.g. naive or concurrent)
+     * @param outputTreeHasher the output tree hasher (e.g. naive or concurrent)
+     */
+    protected AbstractBlockVerificationSession(
+            @NonNull final BlockHeader blockHeader,
+            @NonNull final MetricsService metricsService,
+            @NonNull final SignatureVerifier signatureVerifier,
+            @NonNull final StreamingTreeHasher inputTreeHasher,
+            @NonNull final StreamingTreeHasher outputTreeHasher) {
+        this.blockNumber = blockHeader.number();
+        this.metricsService = metricsService;
+        this.signatureVerifier = signatureVerifier;
+        this.inputTreeHasher = inputTreeHasher;
+        this.outputTreeHasher = outputTreeHasher;
+
+        this.blockWorkStartTime = System.nanoTime();
```

add non null on 98-101 as they are assignment only, lets fail early here

---

✅ Done!